dns service client: 4xx errors may not be handled (all of them, always?) -- regression around `e instanceof got.RequestError`

https://buildkite.com/opstrace/upgrade-tests/builds/1259#de149bfb-607a-4fb0-80a7-77acdd3a7740/8243-9523

```
[2021-08-24T21:48:06Z] 2021-08-24T21:48:06.367Z [32minfo[39m: DNS service teardown: NS record query attempt 1: NS record set exists for upgrade-bk-1259-cb7-g.opstrace.io on the *.opstrace.io DNS zone.
[2021-08-24T21:48:06Z] 2021-08-24T21:48:06.367Z [34mdebug[39m: do DNS service client:DELETE
[2021-08-24T21:48:06Z] 2021-08-24T21:48:06.443Z [31merror[39m: error during cluster teardown (attempt 5):
[2021-08-24T21:48:06Z] HTTPError: Response code 404 (Not Found)
[2021-08-24T21:48:06Z]     at Request.<anonymous> (/snapshot/build/lib/utils/node_modules/got/dist/source/as-promise/index.js:117:42)
[2021-08-24T21:48:06Z]     at runMicrotasks (<anonymous>)
[2021-08-24T21:48:06Z]     at processTicksAndRejections (node:internal/process/task_queues:96:5) {
[2021-08-24T21:48:06Z]   code: undefined,
[2021-08-24T21:48:06Z]   timings: [Object]
[2021-08-24T21:48:06Z] }
[2021-08-24T21:48:06Z] 2021-08-24T21:48:06.443Z [34mdebug[39m: `err?.stack`: HTTPError: Response code 404 (Not Found)
[2021-08-24T21:48:06Z]     at Request.<anonymous> (/snapshot/build/lib/utils/node_modules/got/dist/source/as-promise/index.js:117:42)
[2021-08-24T21:48:06Z]     at runMicrotasks (<anonymous>)
[2021-08-24T21:48:06Z]     at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2021-08-24T21:48:06Z] 2021-08-24T21:48:06.443Z [31merror[39m: JSON representation of err: {
[2021-08-24T21:48:06Z]   "name": "HTTPError",
[2021-08-24T21:48:06Z]   "timings": {
[2021-08-24T21:48:06Z]     "start": 1629841686368,
[2021-08-24T21:48:06Z]     "socket": 1629841686368,
[2021-08-24T21:48:06Z]     "lookup": 1629841686368,
[2021-08-24T21:48:06Z]     "connect": 1629841686376,
[2021-08-24T21:48:06Z]     "secureConnect": 1629841686389,
[2021-08-24T21:48:06Z]     "upload": 1629841686389,
[2021-08-24T21:48:06Z]     "response": 1629841686443,
[2021-08-24T21:48:06Z]     "end": 1629841686443,
[2021-08-24T21:48:06Z]     "phases": {
[2021-08-24T21:48:06Z]       "wait": 0,
[2021-08-24T21:48:06Z]       "dns": 0,
[2021-08-24T21:48:06Z]       "tcp": 8,
[2021-08-24T21:48:06Z]       "tls": 13,
[2021-08-24T21:48:06Z]       "request": 0,
[2021-08-24T21:48:06Z]       "firstByte": 54,
[2021-08-24T21:48:06Z]       "download": 0,
[2021-08-24T21:48:06Z]       "total": 75
[2021-08-24T21:48:06Z]     }
[2021-08-24T21:48:06Z]   }
[2021-08-24T21:48:06Z] }
[2021-08-24T21:48:06Z] 2021-08-24T21:48:06.444Z [31merror[39m: 5 attempt(s) failed. Stop retrying. Exit.
[2021-08-24T21:48:06Z] 2021-08-24T21:48:06.444Z [34mdebug[39m: shut down logger, then exit with code 1
[2021-08-24T21:48:06Z] make[1]: *** [Makefile:760: testupgrade-teardown] Error 1
[2021-08-24T21:48:06Z] make[1]: Leaving directory '/build'
```

--- a/lib/dns/src/opstrace.ts
+++ b/lib/dns/src/opstrace.ts
@@ -40,8 +40,8 @@ const ID_TOKEN_FILE_PATH = "./id.jwt";
 
 // Use e.g. https://httpbin.org/status/500
 // for testing behavior upon 5xx.
-//const DNS_SERVICE_URL = "https://dns-api.opstrace.net/dns/";
-const DNS_SERVICE_URL = "http://httpbin.org/status/404";
+const DNS_SERVICE_URL = "https://dns-api.opstrace.net/dns/";
+//const DNS_SERVICE_URL = "http://httpbin.org/status/401";
 
 const OIDC_ISSUER = "https://opstrace-dev.us.auth0.com";
 const OIDC_CLIENT_ID = "fT9EPILybLT44hQl2xE7hK0eTuH1sb21";
@@ -389,9 +389,14 @@ export class DNSClient {
           // It's actually unlikely that a high-level retry will heal something
           // at this point, but maybe it does. Let's see, maybe it's better UX to
           // `die()`, here, too.
+          log.info(
+            "dns api: not an expected http error response, throw HighLevelRetry"
+          );
+
           throw new HighLevelRetry(`${action} failed`);
         }
 
+        log.info(`http err handler, throw: ${e.name} -- ${e.message}`);
         // Re-throw all other errors, such as JSON.parse() errors.
         throw e;
       }

--- a/lib/dns/src/opstrace.ts
+++ b/lib/dns/src/opstrace.ts
@@ -40,7 +40,8 @@ const ID_TOKEN_FILE_PATH = "./id.jwt";
 
 // Use e.g. https://httpbin.org/status/500
 // for testing behavior upon 5xx.
-const DNS_SERVICE_URL = "https://dns-api.opstrace.net/dns/";
+//const DNS_SERVICE_URL = "https://dns-api.opstrace.net/dns/";
+const DNS_SERVICE_URL = "http://httpbin.org/status/404";
 
 const OIDC_ISSUER = "https://opstrace-dev.us.auth0.com";
 const OIDC_CLIENT_ID = "fT9EPILybLT44hQl2xE7hK0eTuH1sb21";
@@ -64,6 +65,7 @@ async function loginPollRequest(
     },
     // Some HTTP error responses are expected. Do this handling work manually.
     throwHttpErrors: false
+    //responseType: "text"
   };
 
   const url = `${OIDC_ISSUER}/oauth/token`;
@@ -336,7 +338,9 @@ export class DNSClient {
         // Good response, no data in response.
         return undefined;
       } catch (e: any) {
-        if (e instanceof got.RequestError) {
+        // got.RequestError does not always work, therefore also
+        // make the HTTPError name check -- see issue #1323
+        if (e instanceof got.RequestError || e.name === "HTTPError") {
           debugLogHTTPResponse(e.response);
 
           // In the future, a DELETE may result in a 404 when the cluster isn't

